### ISSUE TITLE
Lints in builtin derives are not attributed to the correct AST nodes

### ISSUE THREAD

**Issue by clubby789** [rust-lang/rust#121171]
I tried this code:

```rust
#[deny(byte_slice_in_packed_struct_with_derive)]
#[repr(packed)]
#[derive(Hash)]
struct FlexZeroSlice {
    width: u8,
    data: [u8],
}
```

I expected to see this happen: The code does not compile, as the lint is triggered on the struct.

Instead, this happened: The code compiles with a warning, as the lint uses the crate-root NodeId for reporting.

(This lint specifically uses `ast::CRATE_NODE_ID` rather than `ecx.current_expansion.lint_node_id`, but the value of `lint_node_id` during builtin derive expansion is always `ast::CRATE_NODE_ID`)

Found while implementing a lint for #121053.
